### PR TITLE
introduce DBWriter interface

### PR DESCRIPTION
**Pull request by coocood**
So we can optionally enable raft.

### CODE REVIEW

**Comment by coocood** [comment]
@zhangjinpeng1987 PTAL

File: tikv/mvcc.go [hunk 1]
```diff
@@ -108,15 +125,40 @@ func (store *MVCCStore) Close() error {
...
+	it := store.lockStore.NewIterator()
+	hdrBuf := make([]byte, 8)
+	hdr := (*lockEntryHdr)(unsafe.Pointer(&hdrBuf[0]))
+	for it.SeekToFirst(); it.Valid(); it.Next() {
+		hdr.keyLen = uint32(len(it.Key()))
+		hdr.valLen = uint32(len(it.Value()))
+		writer.Write(hdrBuf)
+		writer.Write(it.Key())
+		writer.Write(it.Value())
+		cnt++
```
Comment (zhangjinpeng1987) [inline tikv/mvcc.go]: Seems cnt not used?

Comment (coocood) [reply tikv/mvcc.go]: I didn't modify the code, just move around, maybe later we can use it to debug.

File: tikv/write.go [hunk 2]
```diff
@@ -214,114 +177,212 @@ func (w *writeLockWorker) run() {
...
+
+func (writer *dbWriter) Close() {
+	close(writer.closeCh)
+	writer.wg.Wait()
+}
+
+func (writer *dbWriter) Write(batch mvcc.WriteBatch) error {
+	wb := batch.(*writeBatch)
+	if len(wb.dbBatch.entries) > 0 {
+		// We must delete lock after commit succeed, or there will be inconsistency.
```
Comment (zhangjinpeng1987) [inline tikv/write.go]: I think this line comment can be moved to L275?

Comment (coocood) [reply tikv/write.go]: ok

File: tikv/write.go [hunk 3]
```diff
@@ -214,114 +177,212 @@ func (w *writeLockWorker) run() {
...
+		atomic.CompareAndSwapUint64(&writer.latestTS, latestTS, ts)
+	}
+}
+
+const delRangeBatchSize = 4096
+
+func (writer *dbWriter) DeleteRange(startKey, endKey []byte, latchHandle mvcc.LatchHandle) error {
+	keys := make([][]byte, 0, delRangeBatchSize)
+	oldStartKey := mvcc.EncodeOldKey(startKey, maxSystemTS)
+	oldEndKey := mvcc.EncodeOldKey(endKey, maxSystemTS)
```
Comment (zhangjinpeng1987) [inline tikv/write.go]: If the range is [start, end), i think you don't need to calculate oldEndKey, just use endkey for old cf.

File: tikv/write.go [hunk 4]
```diff
@@ -214,114 +177,212 @@ func (w *writeLockWorker) run() {
...
-		if err != nil {
-			return errors.Trace(err)
-		}
-		if cap(keyBuf) < int(hdr.keyLen) {
-			keyBuf = make([]byte, hdr.keyLen)
-		}
-		if cap(valBuf) < int(hdr.valLen) {
-			valBuf = make([]byte, hdr.valLen)
+		keys = append(keys, key)
+		if len(keys) == delRangeBatchSize {
```
Comment (zhangjinpeng1987) [inline tikv/write.go]: So the keys just part of the result? Should involve multiple `collectRangeKyes` and `deleteKeysInBatch` at L346?

File: tikv/write.go (hunk 3 above)
Comment (coocood) [reply tikv/write.go]: But `endKey` has prefix 't', `oldStartKey` and `oldEndKey` has prefix `u`.

**Pull request merged**